
code from https://gitlab.collabora.com/detlev/linux/-/commits/add-rkvdec2-driver-iommu

From 96e077ca9e488c96f4cdfa6c23747a4785b49e50 Mon Sep 17 00:00:00 2001
From: Detlev Casanova <detlev.casanova@collabora.com>
Date: Thu, 27 Jun 2024 12:39:28 -0400
Subject: [PATCH] media: rkvdec2: Add iommu support

This allows using all available memory.

The IOMMU is part of the decoder core but can be driven by the rockchip
iommu driver as a separate device.

The only issue with this is that the IOMMU hardware is reset when the
decoder is reset because of an error.
In this case, the IOMMU hardware looses all mappings but the mapped buffer
still retain their mapped address, cached in the IOMMU and decoder drivers.

To avoid that, when a decoding error occurs, the iommu_flush_iotlb_all
callback is used on the IOMMU to reconfigure the HW before continuing.

Signed-off-by: Detlev Casanova <detlev.casanova@collabora.com>
---
 arch/arm64/boot/dts/rockchip/rk3588-base.dtsi | 22 +++++

diff --git a/arch/arm64/boot/dts/rockchip/rk3588-base.dtsi b/arch/arm64/boot/dts/rockchip/rk3588-base.dtsi
index 5c5a186f59a09..8075b0bcf1d97 100644
--- a/arch/arm64/boot/dts/rockchip/rk3588-base.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3588-base.dtsi
@@ -1253,9 +1253,20 @@ vdec0: video-decoder@fdc38000 {
 			 <&cru SRST_RKVDEC0_CORE>, <&cru SRST_RKVDEC0_HEVC_CA>;
 		reset-names = "axi", "ahb", "cabac", "core", "hevc_cabac";
 		power-domains = <&power RK3588_PD_RKVDEC0>;
+		iommus = <&vdec0_mmu>;
 		sram = <&vdec0_sram>;
 	};
 
+	vdec0_mmu: iommu@fdc38700 {
+		compatible = "rockchip,rk3588-iommu", "rockchip,rk3568-iommu";
+		reg = <0x0 0xfdc38700 0x0 0x40>, <0x0 0xfdc38740 0x0 0x40>;
+		interrupts = <GIC_SPI 96 IRQ_TYPE_LEVEL_HIGH 0>;
+		clocks = <&cru ACLK_RKVDEC0>, <&cru HCLK_RKVDEC0>;
+		clock-names = "aclk", "iface";
+		#iommu-cells = <0>;
+		power-domains = <&power RK3588_PD_RKVDEC0>;
+	};
+
 	vdec1: video-decoder@fdc40000 {
 		compatible = "rockchip,rk3588-vdec";
 		reg = <0x0 0xfdc40000 0x0 0x100>, <0x0 0xfdc40100 0x0 0x500>, <0x0 0xfdc40600 0x0 0x100>;
@@ -1272,9 +1283,20 @@ vdec1: video-decoder@fdc40000 {
 			 <&cru SRST_RKVDEC1_CORE>, <&cru SRST_RKVDEC1_HEVC_CA>;
 		reset-names = "axi", "ahb", "cabac", "core", "hevc_cabac";
 		power-domains = <&power RK3588_PD_RKVDEC1>;
+		iommus = <&vdec1_mmu>;
 		sram = <&vdec1_sram>;
 	};
 
+	vdec1_mmu: iommu@fdc40700 {
+		compatible = "rockchip,rk3588-iommu", "rockchip,rk3568-iommu";
+		reg = <0x0 0xfdc40700 0x0 0x40>, <0x0 0xfdc40740 0x0 0x40>;
+		interrupts = <GIC_SPI 96 IRQ_TYPE_LEVEL_HIGH 0>;
+		clocks = <&cru ACLK_RKVDEC0>, <&cru HCLK_RKVDEC0>;
+		clock-names = "aclk", "iface";
+		#iommu-cells = <0>;
+		power-domains = <&power RK3588_PD_RKVDEC0>;
+	};
+
 	av1d: video-codec@fdc70000 {
 		compatible = "rockchip,rk3588-av1-vpu";
 		reg = <0x0 0xfdc70000 0x0 0x800>;
